### PR TITLE
GSdx Merge Circuit: Fix regression and issue

### PR DESCRIPTION
**Pull request by FlatOutPS2**
EDIT: Fixes graphical issues in NASCAR 09 and reverses changes that caused a regression in Time Crisis 2/3 split screen mode.

https://github.com/PCSX2/pcsx2/pull/1683 should prevent the Software mode crashes in the EA NASCAR games.

### CODE REVIEW

**Comment by willkuer** [comment]
Was it finally an uint/int issue?

I think the code deserves comments regarding timesplitters and nascar. Especially the line //shouldn't it be >2 that you changed from 4 to 40 deserves an update (i.e. why 40?)

**Comment by FlatOutPS2** [comment]
> Was it finally an uint/int issue?

Yeah. The value (2047) made no sense to me otherwise, so I gave this a try and it immediately resolved the issue.

> I think the code deserves comments regarding timesplitters and nascar. Especially the line //shouldn't it be >2 that you changed from 4 to 40 deserves an update (i.e. why 40?)

That was a mistake. :p The github app doesn't show changes as clear as the website does.

**Comment by ssakash** [comment]
Code doesn't make sense to me,  DBX/DBY is a bit field with 11 bit width and no sign bit. So not sure why you converted it to ``int``?

**Comment by gregory38** [comment]
What is DBX/DBY again ? Position offset.

GS memory wrap after 4MB, you must restart to 0. If code is wrong, it will overflow the memory and then crash. (as a trick GS memory is bigger to avoid small overflow but too big overflow will explode).

**Comment by gregory38** [comment]
Is there any gs dump that can reproduce the crash ?

**Comment by ssakash** [comment]
Yes, a GS dump of the crash would help. The crash is most likely caused due to either a bug in software mode (or) an insane register value init by the developer on ``DISPFB``, I hope it's the former.

**Comment by FlatOutPS2** [comment]
> The crash might most likely be caused due to either a bug in software mode (or) an insane register value init by the developer on DISPFB, I hope it's the former.

Do you consider setting DBY to -1 an insane value?

**Comment by ssakash** [comment]
> Do you consider setting DBY to -1 an insane value?

Insane? No, it's an impossible value. IIRC the data type format of DBX/DBY on the Graphics synthesizer is ``int 0:11:0`` which means no negative values as there's no sign bit.

**Comment by FlatOutPS2** [comment]
> Insane? No, it's an impossible value. IIRC the data type format of DBX/DBY on the Graphics synthesizer is int 0:11:0 which means no negative values as there's no sign bit.

And because it's impossible it gets converted to 2047.

**Comment by ssakash** [comment]
> And because it's impossible it gets converted to 2047.

I'm not sure, if the developer did indeed pass ``-1`` then it would most likely be stored as ``10000000001`` which would then be converted into``1025`` with respect to the integer format of DBY.

**Comment by FlatOutPS2** [comment]
> I'm not sure, if the developer did indeed pass -1 then it would most likely be stored as 10000000001 which would then be converted into1025 with respect to the integer format of DBY.

When I set the value to -1 manually it got converted to 2047 as well.

**Comment by ssakash** [comment]
> When I set the value to -1 manually it got converted to 2047 as well.

Oh, then it probably stores the 2's complement value. However, I still don't think any developer would be stupid enough to assign an illegal value to the bit-fields.

**Comment by gregory38** [comment]
Framebuffer size is 2048x2048, so 2047 seem valid. And the memory wraps after 4MB. There are nothing illegal on the gs.

**Comment by gregory38** [comment]
Any backtrace on the crash ? I'm pretty sure that we miss some memory wrapping.

File: plugins/GSdx/GSRenderer.cpp [hunk 1]
```diff
@@ -139,59 +139,27 @@ bool GSRenderer::Merge(int field)
...
-			// dr[1] = 127 50 639 494
-
-			int top = min(fr[0].top, fr[1].top);
-			int bottom = min(fr[0].bottom, fr[1].bottom);
-
-			fr[0].top = fr[1].top = top;
-			fr[0].bottom = fr[1].bottom = bottom;
-
-			// blurdetected = true;
+			if (dr[0].eq(dr[1]) && (fr[0].eq(fr[1] + GSVector4i(0, j, 0, j)) || fr[1].eq(fr[0] + GSVector4i(0, j, 0, j))))
```
Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: There's got to be a better way than using a for loop. Maybe something along the lines of:
```
diff = fr[0] - fr[1];
if (dr[0].eq(dr[1]) && diff.x == diff.z && diff.x == 0 && diff.y == diff.w)
{
    // do whatever
```
Well, I think that's what your change is equivalent to.

Comment (FlatOutPS2) [inline plugins/GSdx/GSRenderer.cpp]: > There's got to be a better way than using a for loop.

I think my thinking got stuck inside of it...

Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: You need to fix the indentation of your new commit BTW.

Comment (FlatOutPS2) [inline plugins/GSdx/GSRenderer.cpp]: Late night text editor changes; not recommended...

File: plugins/GSdx/GSRenderer.cpp [hunk 2]
```diff
@@ -237,18 +202,17 @@ bool GSRenderer::Merge(int field)
...
 		}
 		// If the DX offset is too small then consider the status of frame memory offsets, prevents blurring on Tenchu: Fatal Shadows, Worms 3D
-		else if((display_diff.x != frame_diff.x) && display_diff.x < 10 && frame_diff.x < 10)
+		else if(display_diff.x != frame_diff.x)
 		{
 			off.x = tex[i]->GetScale().x * frame_diff.x;
 		}
 
-		if(display_diff.y >= 4 && display_diff.y < 10) // Shouldn't this be >= 2?
+		if(display_diff.y >= 4) // Shouldn't this be >= 2?
```
Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: > // Shouldn't this be >= 2?

Maybe you could update this comment explaining why it's ``>=4`` instead?

File: plugins/GSdx/GSRenderer.cpp [hunk 3]
```diff
@@ -139,42 +139,10 @@ bool GSRenderer::Merge(int field)
...
-			fr[0].top = top;
-			fr[1].top = top;
-			dr[0].bottom = bottom;
-			dr[1].bottom = bottom;
-
-			// blurdetected = true;
-		}
-		else if(dr[0].eq(dr[1]) && (fr[0].eq(fr[1] + GSVector4i(0, 1, 0, 1)) || fr[1].eq(fr[0] + GSVector4i(0, 1, 0, 1))))
+		int topDiff = fr[0].top - fr[1].top;
+		if (dr[0].eq(dr[1]) && (fr[0].eq(fr[1] + GSVector4i(0, topDiff, 0, topDiff)) || fr[1].eq(fr[0] + GSVector4i(0, topDiff, 0, topDiff))))
```
Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: > dr[0].eq(dr[1])

What about games which differ on the output circuit? God of war and Persona 4 for example as pointed out in the code comments.

```
            // persona 4:
			//
			// fr[0] = 0 0 640 448
			// fr[1] = 0 1 640 448
			// dr[0] = 159 50 779 498
			// dr[1] = 159 50 779 497
			//
			// second image shifted up by 1 pixel and blended over itself
			//
			// god of war:
			//
		    // fr[0] = 0 1 512 448
		    // fr[1] = 0 0 512 448
		    // dr[0] = 127 50 639 497
     	    // dr[1] = 127 50 639 498
		    //

Comment (FlatOutPS2) [inline plugins/GSdx/GSRenderer.cpp]: These cases are fixed by the display_diff/frame_diff code below, so it's become obsolete.

File: plugins/GSdx/GSRenderer.cpp (hunk 2 above)
Comment (FlatOutPS2) [inline plugins/GSdx/GSRenderer.cpp]: And why is it `>=4` ? :P

Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: No clue, you were the one who added it :P

Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: Nevermind, it seems it has been 4 even before the 1.4 times.

File: plugins/GSdx/GSRenderer.cpp (hunk 3 above)
Comment (gregory38) [inline plugins/GSdx/GSRenderer.cpp]: Could you keep the comment for history? Saying that is obsolete and that is fixed with code blabla..

Comment (FlatOutPS2) [inline plugins/GSdx/GSRenderer.cpp]: > Could you keep the comment for history? Saying that is obsolete and that is fixed with code blabla..

The entire code or just the Persona and GOW offset details?

Note that many games actually hit this code, not just those two games.

**Comment by refractionpcsx2** [comment]
Well I've thrown some games at this and all seems well as normal :)

You okay with it @gregory38 as I see you weren't sure on something?

**Comment by gregory38** [comment]
Fine PR was updated. You can merge it

**Comment by willkuer** [comment]
This pr seems to introduce a regression in Warship Gunner 2 (#1723). Seems like depth handling broke.

**Comment by refractionpcsx2** [comment]
Are you sure it's this commit? Those depth issues are usually issues involved with VIF and VU timing.

Nice timing btw :P Might have been more useful before the merge lol

**Comment by disastranagant** [comment]
The recent VIF commits were why I was testing it but all recent builds but this one work.

**Comment by refractionpcsx2** [comment]
Fair enough, didn't think this commit would cause that sort of problem.  If you can do a GS Dump, maybe @FlatOutPS2 would like to look in to it.

**Comment by disastranagant** [comment]
dump https://www.dropbox.com/s/z8leoubpc3hrksr/warship%20gunner%202%20regression.7z?dl=0

**Comment by FlatOutPS2** [comment]
Thanks for the dump. It has nothing to do with this PR. It's depth issue, not a PCRTC issue, so it would've been very odd for this to trigger it. See https://github.com/PCSX2/pcsx2/issues/1723 for details.

**Comment by MrCK1** [comment]
Nascar 09 is all good - still not sure why it leaks memory though

**Pull request merged**